contracts/Chef.sol

```diff
@@ -0,0 +1,198 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.0;
+
+import "hardhat/console.sol";
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import "./MetricToken.sol";
+
+// Heavily Inspired by Sushi's MasterChefv2 - but with a few changes:
+// - We don't have a v1, so we don't need that wrapping
+// - We don't have two layers (pools and users), so the concept of pools is flattened into the contract itself.
+// ^^ This is because METRIC is the only token this will ever work with.
+
+// Read this: https://dev.sushi.com/sushiswap/contracts/masterchefv2
+// Also read this: https://soliditydeveloper.com/sushi-swap
+
+/**
+ In sushi's master chef, the design allows the controller to submit LP contracts for yield farming, and then user's can stake their LP tokens.
+
+ In this contract, there is no concept of a user staking their LP tokens - and instead of LP contract, the controller is submitting Allocation Groups.
+
+ So in sushi:
+
+ 1.  Every `x` blocks, calculate minted Sushi Tokens for each LP contract based on their (shares / total shares)
+ 2.  Then, do the math to figure out how many rewards each LP token is worth (based on the total amount of LP tokens staked)
```

Comment:
Why is the Sushi algorithm being described here? Feels like good info for the PR description but not needed to document _this_ code?

Reply:
this is pretty common with smart contracts because the code and its comments will be available on etherscan (or the chain's fork).  I also think it's pretty web3-style because it means the comments are decentralized and immutable as well as the code.

I will pull it into the readme, and I expect us to have a gitbook at some point where it can also live.  Once our documentation source is finalized (assuming before launch) we can link to that from the smart contract

Reply:
The code being commented is fine, I'm confused why the comments are describing _other_ code. Fine if you're pulling that out, I'd just expect "We started with this other thing someone else wrote, so if you're interested you can look there, but we here we describe our thing"

Reply:
hmmm.  Good point!  the original intention was that I had to wrap my head around sushi's masterchef completely to be able to modify it so I could write this one - so I was partially writing those comments to help my understanding.

But now, we have our own implementation so this does make sense.